File: scripts/pty-server.py
#!/usr/bin/env python3
"""
WebSocket PTY server for claude auth login.
Runs on port 3100. FE connects via xterm.js + WebSocket.

Usage: python3 scripts/pty-server.py
"""
import asyncio
import json
import os
import pty
import select
import signal
import struct
import subprocess
import sys
import fcntl
import termios

try:
    import websockets
    from websockets.asyncio.server import serve
except ImportError:
    print("Installing websockets...")
    subprocess.check_call([sys.executable, "-m", "pip", "install", "websockets", "-q"])
    import websockets
    from websockets.asyncio.server import serve

PORT = 3100

def get_claude_binary():
    try:
        return subprocess.check_output(["which", "claude"], text=True, timeout=3).strip()
    except:
        return "claude"

def get_auth_status():
    try:
        binary = get_claude_binary()
        out = subprocess.check_output([binary, "auth", "status"], text=True, timeout=10, stderr=subprocess.STDOUT).strip()
        data = json.loads(out)
        return {"loggedIn": data.get("loggedIn", False), "email": data.get("email")}
    except:
        return {"loggedIn": False}

async def handle_client(websocket):
    path = websocket.request.path if hasattr(websocket, 'request') else "/"
    action = "login"
    if "action=status" in str(path):
        action = "status"
    elif "action=logout" in str(path):
        action = "logout"

    print(f"[pty-server] client connected, action={action}")

    if action == "status":
        await websocket.send(json.dumps({"type": "status", **get_auth_status()}))
        return

    if action == "logout":
        binary = get_claude_binary()
        try:
            subprocess.run([binary, "auth", "logout"], capture_output=True, timeout=10)
        except:
            pass
        await websocket.send(json.dumps({"type": "logout", "success": True}))
        return

    # Login — spawn PTY
    binary = get_claude_binary()
    master, slave = pty.openpty()

    pid = os.fork()
    if pid == 0:
        # Child
        os.close(master)
        os.setsid()
        os.dup2(slave, 0)
        os.dup2(slave, 1)
        os.dup2(slave, 2)
        os.close(slave)
        os.environ["TERM"] = "xterm-256color"
        os.execvp(binary, [binary, "auth", "login"])
        os._exit(1)

    os.close(slave)
    print(f"[pty-server] spawned claude auth login, pid={pid}")

    # Set master to non-blocking
    flags = fcntl.fcntl(master, fcntl.F_GETFL)
    fcntl.fcntl(master, fcntl.F_SETFL, flags | os.O_NONBLOCK)

    loop = asyncio.get_event_loop()
    running = True

    async def read_pty():
        """Read from PTY and send to WebSocket."""
        while running:
            try:
                r, _, _ = select.select([master], [], [], 0.05)
                if r:
                    data = os.read(master, 4096)
                    if not data:
                        break
                    await websocket.send(data)
                else:
                    await asyncio.sleep(0.05)
            except OSError:
                break
            except websockets.exceptions.ConnectionClosed:
                break

        # Check final status
        status = get_auth_status()
        print(f"[pty-server] auth status: loggedIn={status['loggedIn']}")
        try:
            await websocket.send(json.dumps({"type": "exit", **status}))
        except:
            pass

    async def write_pty():
        """Read from WebSocket and write to PTY."""
        nonlocal running
        try:
            async for message in websocket:
                if isinstance(message, str):
                    # Check for resize
                    try:
                        parsed = json.loads(message)
                        if parsed.get("type") == "resize":
                            cols = parsed.get("cols", 80)
                            rows = parsed.get("rows", 24)
                            winsize = struct.pack("HHHH", rows, cols, 0, 0)
                            fcntl.ioctl(master, termios.TIOCSWINSZ, winsize)
                            continue
                    except (json.JSONDecodeError, ValueError):
                        pass
                    os.write(master, message.encode())
                else:
                    os.write(master, message)
        except websockets.exceptions.ConnectionClosed:
            pass
        finally:
            running = False

    try:
        await asyncio.gather(read_pty(), write_pty())
    finally:
        print(f"[pty-server] cleaning up pid={pid}")
        try:
            os.close(master)
        except:
            pass
        try:
            os.kill(pid, signal.SIGTERM)
        except:
            pass
        try:
            os.waitpid(pid, 0)
        except:
            pass

async def main():
    async with serve(handle_client, "localhost", PORT):
        print(f"[pty-server] listening on ws://localhost:{PORT}")
        await asyncio.Future()  # run forever

if __name__ == "__main__":
    asyncio.run(main())

File: scripts/pty-server.ts
/**
 * Standalone WebSocket PTY server for Claude interactions.
 * Runs on port 3100. The FE connects via xterm.js + WebSocket.
 *
 * Actions:
 *   ?action=login           — interactive `claude auth login`
 *   ?action=status           — auth status check (returns JSON, closes)
 *   ?action=logout           — auth logout (returns JSON, closes)
 *   ?action=execute&cwd=...  — interactive Claude session in a directory
 *
 * Usage: npx tsx scripts/pty-server.ts
 */
import { WebSocketServer, WebSocket } from "ws";
import * as pty from "node-pty";
import { execSync } from "child_process";

const PORT = 3100;

function getClaudeBinary(): string {
  try {
    return execSync("which claude 2>/dev/null", { encoding: "utf-8", timeout: 3000 }).trim();
  } catch {
    return "claude";
  }
}

function getAuthStatus(): { loggedIn: boolean; email?: string } {
  try {
    const binary = getClaudeBinary();
    const out = execSync(`${binary} auth status 2>&1`, { encoding: "utf-8", timeout: 10000 }).trim();
    const data = JSON.parse(out);
    return { loggedIn: data.loggedIn === true, email: data.email };
  } catch {
    return { loggedIn: false };
  }
}

const wss = new WebSocketServer({ port: PORT });
console.log(`[pty-server] listening on ws://localhost:${PORT}`);

wss.on("connection", (ws: WebSocket, req) => {
  const url = new URL(req.url || "/", `http://localhost:${PORT}`);
  const action = url.searchParams.get("action") || "login";
  const binary = getClaudeBinary();
  console.log(`[pty-server] client connected, action=${action}, binary=${binary}`);

  // --- STATUS ---
  if (action === "status") {
    const status = getAuthStatus();
    ws.send(JSON.stringify({ type: "status", ...status }));
    ws.close();
    return;
  }

  // --- LOGOUT ---
  if (action === "logout") {
    try {
      execSync(`${binary} auth logout 2>&1`, { encoding: "utf-8", timeout: 10000 });
    } catch { /* ok */ }
    ws.send(JSON.stringify({ type: "logout", success: true }));
    ws.close();
    return;
  }

  // --- LOGIN --- spawn PTY for `claude auth login`
  if (action === "login") {
    const shell = pty.spawn(binary, ["auth", "login"], {
      name: "xterm-256color",
      cols: 100,
      rows: 24,
      env: { ...process.env, TERM: "xterm-256color" } as Record<string, string>,
    });

    console.log(`[pty-server] spawned claude auth login, pid=${shell.pid}`);
    wireUpPty(ws, shell, () => {
      const status = getAuthStatus();
      console.log(`[pty-server] auth status: loggedIn=${status.loggedIn}`);
      return { type: "exit", ...status };
    });
    return;
  }

  // --- EXECUTE --- spawn interactive Claude session in a working directory
  if (action === "execute") {
    const cwd = url.searchParams.get("cwd") || process.cwd();
    console.log(`[pty-server] execute in cwd=${cwd}`);

    const shell = pty.spawn(binary, [], {
      name: "xterm-256color",
      cols: 100,
      rows: 30,
      cwd,
      env: { ...process.env, TERM: "xterm-256color" } as Record<string, string>,
    });

    console.log(`[pty-server] spawned interactive claude, pid=${shell.pid}`);
    wireUpPty(ws, shell, () => ({ type: "exit" }));
    return;
  }

  ws.send(JSON.stringify({ type: "error", message: `Unknown action: ${action}` }));
  ws.close();
});

/**
 * Wire a PTY process to a WebSocket connection.
 * Handles bidirectional data, resize, and cleanup.
 */
function wireUpPty(
  ws: WebSocket,
  shell: pty.IPty,
  onExit?: () => Record<string, unknown>
) {
  // PTY → WebSocket
  shell.onData((data: string) => {
    if (ws.readyState === WebSocket.OPEN) {
      ws.send(data);
    }
  });

  // WebSocket → PTY (user keystrokes or control messages)
  ws.on("message", (msg: Buffer | string) => {
    const str = typeof msg === "string" ? msg : msg.toString();
    try {
      const parsed = JSON.parse(str);
      if (parsed.type === "resize") {
        shell.resize(parsed.cols, parsed.rows);
        return;
      }
    } catch {
      // Not JSON — raw keystroke input
    }
    shell.write(str);
  });

  shell.onExit(({ exitCode }) => {
    console.log(`[pty-server] process exited, code=${exitCode}`);
    if (ws.readyState === WebSocket.OPEN) {
      const exitMsg = onExit ? onExit() : { type: "exit" };
      ws.send(JSON.stringify({ ...exitMsg, exitCode }));
      ws.close();
    }
  });

  ws.on("close", () => {
    console.log(`[pty-server] client disconnected`);
    try { shell.kill(); } catch { /* ok */ }
  });
}
